Voice call flow update (#3521)
Nexmo/nexmo-developer: _use_cases/en/contact-center.md

---
title: Contact Center Augmentation
products: voice/voice-api
description: "This tutorial shows you how to add programmable assistance to your contact center"
languages:
    - Node
navigation_weight: 5    
---

# Contact Center Augmentation

In this tutorial, you will implement the following contact center scenario:

> * the user calls the phone number,

Reviewer: "the phone number" or "your Vonage number"? Please capitalise first letter of bullets

> * initial greeting is played, and the user selects an option with IVR,

Reviewer: ```suggestion
* The initial greeting is read out, and the user chooses an option from the menu provided
```

const bodyParser = require('body-parser')
const https = require('https')
const Vonage = require('@vonage/server-sdk')

const vonage = new Vonage({
  apiKey: <YOUR_API_KEY>,
  apiSecret: <YOUR_API_SECRET>,
  applicationId: <YOUR_APP_ID>,
  privateKey: 'private.key'
})

app.use(bodyParser.json())

app.get('/webhooks/answer', (request, response) => {
  console.log('answer: ', request.query)

  const ncco = [{
    action: 'talk',
    text: 'Thank you for calling Example Inc.! Press 1 to talk to the sales department, press 2 to get technical support.',
    bargeIn: true
  },
    {
      action: 'input',
      eventUrl: [
        `${request.protocol}://${request.get('host')}/webhooks/input`],
      type: [ 'dtmf' ],
      dtmf: {
        maxDigits: 1
      }
    }
  ]

  response.json(ncco)
})
```

## Write your event webhook

Implement a webhook that captures call events so that you can observe the lifecycle of the call in the console:

```js
app.post('/webhooks/event', (request, response) => {
  console.log('event:', request.body)
  response.sendStatus(200);
})
```

Vonage makes a `POST` request to this endpoint every time the call status changes.

## Write your input webhook

> DTMF input results will be sent to the specific URL you set in the `input` action: `/webhooks/input`. Add a webhook to process the result and add some user interaction.

Reviewer: ```suggestion
DTMF input will be sent to the specific URL you set in the `input` action: `/webhooks/input`. Add a webhook to process the result and provide some user interaction.
```

In case of a successful recognition, the request payload will look as follows:

```json
{
  "speech": {
  },
  "dtmf": {
    "digits": "1",
    "timed_out": true
  },

>   "from": <USER_NUMBER>,

Reviewer: We should be using standard variable placeholder names here - see the `/config/code_snippet_variables.yml` file

Author: I couldn't find a proper standard name for this... could you please advise?

  "to": <LVN>,
  "uuid": "abfd679701d7f810a0a9a44f8e298b33",
  "conversation_uuid": "CON-64e6c8ef-91a9-4a21-b664-b00a1f41340f",
  "timestamp": "2020-04-17T17:31:53.638Z"
}
```

> So you should use the `dtmf.digits` array for further analysis. To connect both user and agent of the corresponding department, you'll create an outbound call to the agent endpoint. You can use a test phone number in this sample for simplicity, for example, your mobile number. In real life case, the endpoint might be still a PSTN number or [SIP endpoint](/voice/sip/concepts/programmable-sip) to connect your existing PBX or contact center, or WebRTC client if you're developing your contact center solution from scratch - Vonage [Client SDK](/client-sdk/overview) provides all the features required for implementing your own Contact Center agent application.

Reviewer: ```suggestion
Retrieve the user's input from the `dtmf.digits` array to determine the options they selected. To connect both user and agent of the corresponding department, you'll create an outbound call to the agent endpoint. You can use a test phone number in this sample for simplicity, for example, your mobile number. In real life, the endpoint might be a PSTN number or [SIP endpoint](/voice/sip/concepts/programmable-sip) to connect your existing PBX or contact center. It could also be a WebRTC client if you are developing your contact center solution from scratch - the Vonage [Client SDK](/client-sdk/overview) provides everything you need to implement your own Contact Center agent application.
```

> Finally, both calls (legs) should be moved to one conference room (named [conversation](/voice/voice-api/ncco-reference#conversation)). To do that, you should use `conversation` action with the same `name` both for the user call (the inbound leg) and the agent leg (the outbound call). To generate the conversation name, you may use any unique ID generation method, for example, using the actual timestamp.

Reviewer: ```suggestion
Finally, both calls (legs) should be moved to one conference room (named [conversation](/voice/voice-api/ncco-reference#conversation)). To do that, use the `conversation` action with the same `name` both for the user call (the inbound leg) and the agent call (the outbound leg). To generate the conversation name, you may use any unique ID generation method (for example, a timestamp).
```

> > As an option, you may use `connect` action in the NCCO to connect the user to the agent. The difference is that with `connect`, the call will be immediately completed when any of the call participants hang up, and there is only one leg left. So, it would be impossible to transfer the user to the survey after the call; if that's not needed in your case, `connect` is a bit more handy option. If you want the user still connected after the agent completes the call, choose `conversation` as shown in the example below.

Reviewer: ```suggestion
> As an option, you may use `connect` action in the NCCO to connect the user to the agent. The difference is that with `connect`, the call will terminate as soon as any participant hangs up, and there is only one leg left. If that happens, it would be impossible to transfer the user to the survey after the call. If you want the user to remain connected to the call after the agent leaves, choose the `conversation` action instead, as shown in the example below.
```

Add the code to handle the input callback:

```js
app.post('/webhooks/input', (request, response) => {
  console.log('input:', request.body)

  // generating unique conversation name
  var conversationName = 'conversation_' + Date.now()

  console.log('conversationName: ', conversationName)

  // selecting agent/department endpoint
  var departmentId = request.body.dtmf.digits
  var department = ''
  var departmentNumber = ''

  switch (departmentId) {
    case '1':
      department = 'Sales'

>       departmentNumber = <YOUR_MOBILE_NUMBER>

Reviewer: Please use standard variable names

Author: Can I use something custom like {{AGENT_NUMBER}}? Or could you please advise which of the standard ones can be used here?

  }

  var ncco = ''

  if (department != '') {
      // NCCO for the user leg
    ncco = [{
      action: 'talk',
      text: 'Please wait while we connect you to ' + department
    }, {
      action: 'conversation',
      name: conversationName
    }
    ]

    // creating the agent leg and moving it to the same conversation
    vonage.calls.create({
      to: [
        {
          type: 'phone',
          number: departmentNumber
        }
      ],
      from: {
        type: 'phone',
        number: <YOUR_LVN>
      },
      ncco: [
        {
          action: 'conversation',
          name: conversationName
        }]

    }, (error, response) => {
      if (error) console.error('outbound error:', error)
      if (response) {
          console.log('outbound ok')
      }
    })
  } else { // something went wrong, fallback route
    ncco = [{
      action: 'talk',
      text: 'Press 1 to talk to the sales department, press 2 to get technical support.',
      bargeIn: true
    }, {
      action: 'input',
      eventUrl: [
        `${request.protocol}://${request.get('host')}/webhooks/input`
      ],
      dtmf: {
        maxDigits: 1
      }
    }]
  }
```

## Add survey NCCO

> Next, to implement the customer satisfaction survey at the end of the call, you should handle `completed` event for the agent's leg. It will arrive at the same event webhook, so you should extend the event webhook with [transfer request](/voice/voice-api/code-snippets/transfer-a-call-inline-ncco) to the survey NCCO. In order to do that, you have to store the user and the agent leg identifiers:

Reviewer: ```suggestion
Next, to implement the customer satisfaction survey at the end of the call, you should handle the `completed` event for the agent's leg. It will arrive at the same event webhook, so you should extend the event webhook with a [transfer request](/voice/voice-api/code-snippets/transfer-a-call-inline-ncco) to the survey NCCO. To do that, you must store the user and the agent leg identifiers:
```

```
var userLegId = ''
var agentLegId = ''

app.get('/webhooks/answer', (request, response) => {
  console.log('answer: ', request.query)

  userLegId = request.query.uuid

  console.log('userLegId: ', userLegId)
  

...

app.post('/webhooks/input', (request, response) => {
  console.log('input:', request.body)

    // creating the agent's leg and moving it to the same conversation
    vonage.calls.create({
        ...
    }, (error, response) => {
      if (error) console.error('outbound error:', error)
      if (response) {
        agentLegId = response.uuid
        console.log('agentLegId: ', agentLegId)
      }
    })
  } else ...
```

> > In the real-life case, you should implement a cache to store the pairs of user/agent leg identifiers and. The sample code shown in this tutorial will work properly only for one concurrent call.

Reviewer: "In A real-life SCENARIO" (or "situation" or "use case" but not just "case". Also this sentence is incomplete (ends with "and")

> Extend your event webhook with REST API update call method with inline NCCO with `talk` and `input` actions to move the user leg to the survey part:

Reviewer: ```suggestion
Extend your event webhook with a call to the `update` method, supplying an inline NCCO that contains `talk` and `input` actions to direct the user to the survey:
```

```js
app.post('/webhooks/event', (request, response) => {
  console.log('event:', request.body)

  if (request.body.uuid == agentLegId && request.body.status == 'completed') {
    vonage.calls.update(userLegId, {
      action: 'transfer',
      destination: {
        type: 'ncco',
        ncco: [ {
          action: 'talk',
          text: 'Please valuate quality of service by entering a digit, 1 to 5'
        },
        {
          action: 'input',
          type: [ 'dtmf' ],
          dtmf: {
            maxDigits: 1
          },
          eventUrl: [ `${request.protocol}://${request.get('host')}/webhooks/survey` ]
        }
        ]
      }
    }, (err, res) => {
      if (err) {
        console.error('transfer error:', err)
      } else {
        console.log('transfer ok')
      }
    })
  } else {
    response.sendStatus(200)
  }
})
```

## Write your survey webhook

> Add survey webhook to print the results:

Reviewer: ```suggestion
Create the survey webhook to print the results:
```

```js
app.post('/webhooks/survey', (request, response) => {
  console.log('survey: ', request.body)

  var phone = request.body.from
  var date = request.body.timestamp
  var score = request.body.dtmf.digits

  console.log('[%s] User %s gave %d', date, phone, score)

  response.sendStatus(200)
})
```

## Create your Node.js server

Finally, write the code to instantiate your Node.js server:

```js
const port = 3000
app.listen(port, () => console.log(`Listening on port ${port}`))
```

## Test your application

1. Run your Node.js application by executing the following command:

```sh
node index.js
```

2. Call your Vonage number from your mobile phone and listen to the welcome message.

> 3. Open the dial pad and press 1 or 2.

Reviewer: ```suggestion
3. Using your phone's dial pad press either 1 or 2.
```

> 3. Answer the inbound call on your second mobile phone (you have to use at least two devices to get the scenario working; alternatively, you may use some calling application, for example, [Vonage Business](https://www.vonage.com/unified-communications/features) app).

Reviewer: If you need two devices, let's say this right at the beginning of the use case.

Let's also state here which device is the agent, etc so that we can stop referring to them as "first", "second", etc

Author: Added to **Prerequisites**

4. Hang up on the second ("agent") device.

5. Listen to the survey message and press any key on the first device.

6. Observe the console log to see the survey result.

## Troubleshooting

If you don't hear the user and agent sound, potentially it might be because the two legs are being processed in different locations. You can determine this by seeing different `conversation_uuid_to` values in the `transfer` events for user and agent legs. To fix that, try to configure the SDK to use a specific data center as described in the [Troubleshooting guide](/voice/voice-api/guides/troubleshooting/node#regions):

```js
const options = {
  apiHost: 'api-us-1.nexmo.com',
  restHost: 'rest-us-1.nexmo.com'
}

const vonage = new Vonage({
  apiKey: <YOUR_API_KEY>,
  apiSecret: <YOUR_API_SECRET>,
  applicationId: <YOUR_APP_ID>,
  privateKey: 'private.key'
}, options)
```

## Conclusion

> With Vonage Voice API you can empower your existing contact center solution with IVR of any logic complexity, which depends only on your target use case and is virtually unlimited. Or you can build your own solution from scratch using the Voice API and [Client SDK](/client-sdk/overview). Switching between the scripted part of the call and live conversation (and back) gives you the ability to mix any phone calling use cases and create a seamless customer experience.

Reviewer: ```suggestion
With the Vonage Voice API you can empower your existing contact center solution with sophisticated IVR logic that meets the needs of your business and customers. Or you can build your own solution from scratch using the Voice API and [Client SDK](/client-sdk/overview). Switching between the scripted part of the call and live conversation enables you to create a seamless customer experience.
```

> ## Where Next?

Reviewer: Capitalize first letter of bullets please
